pkg/cvo/status_history.go

```diff
@@ -0,0 +1,235 @@
+package cvo
+
+import (
+	"math"
+	"strings"
+
+	"k8s.io/klog/v2"
+
+	configv1 "github.com/openshift/api/config/v1"
+)
+
+const (
+	// maxFinalEntryIndex is an upper bound index. History entries from 0 to maxFinalEntryIndex are given mostImportantWeight.
+	maxFinalEntryIndex = 4
+
+	// mostImportantWeight is given to history entries that should never be removed including the initial entry,
+	// the most recent entries (indices 0..maxFinalEntryIndex), and the most recently completed update.
+	mostImportantWeight = 1000.0
+
+	// interestingWeight is given to the first or last completed update in a given minor. These are interesting
+	// but not critical.
+	interestingWeight = 30.0
+
+	// partialMinorWeight is given to any partial minor updates between a completed minor version transition.
+	partialMinorWeight = 20.0
+
+	// partialZStreamWeight is given to any partial z-stream updates between a completed z-stream version transition.
+	partialZStreamWeight = -20.0
+
+	// sliceIndexWeight when applied will favor more recent updates and avoid ties.
+	sliceIndexWeight = -1.01
+)
+
+// maxHistorySize is set by the function prune and used by all the history pruning functions. Setting and using
+// this local package variable eases unit testing by allowing testing against a smaller history slice.
+var maxHistorySize int
+
+// prune prunes history when at maxSize by ranking each entry using the above defined weights and removing the entry with
+// the lowest rank. maxSize is passed in to allow the use of smaller values which eases unit test.
+func prune(history []configv1.UpdateHistory, maxSize int) []configv1.UpdateHistory {
+	if len(history) <= maxSize {
+		return history
+	}
+	maxHistorySize = maxSize
+
+	idx := getTheMostRecentCompletedEntryIndex(history)
+
+	var reason string
+	lowestRank := math.MaxFloat64
+	var lowestRankIdx int
+
+	for i := range history {
+		rank := 0.0
+		thisReason := ""
+		if isTheInitialEntry(i) {
+			rank = mostImportantWeight
+		}
+		if rank == 0.0 && isAFinalEntry(i) {
+			rank = mostImportantWeight
+		}
+		if rank == 0.0 && isTheMostRecentCompletedEntry(i, idx) {
+			rank = mostImportantWeight
+		}
+		if rank == 0.0 && isTheFirstOrLastCompletedInAMinor(i, history) {
+			rank = rank + interestingWeight
+			thisReason = "the first or last completed minor update"
+		}
+		if rank == 0.0 && isPartialPortionOfMinorTransition(i, history) {
+			rank = rank + partialMinorWeight
+			thisReason = "a partial update within a minor transition"
+		}
+		if rank == 0.0 && isPartialWithinAZStream(i, history) {
+			rank = rank + partialZStreamWeight
+			thisReason = "a partial update within a z-stream transition"
+		}
+		rank = rank + (sliceIndexWeight * float64(i))
+
+		if rank < lowestRank {
+			lowestRank = rank
+			lowestRankIdx = i
+			reason = thisReason
+		}
+	}
+	if reason == "" {
+		klog.V(2).Infof("Pruning version %s at index %d with rank %f.", history[lowestRankIdx].Version, lowestRankIdx, lowestRank)
```

Comment:
nit: maybe we should drop this case, and default `reason := "<bug should never prune without a reason>"` or something?  Because reading the cases above, you're setting `reason` for everything except completed within a z stream, and maybe that deserves an explicit case and a `a completed update within a z-stream transition` `thisReason`?

Reply:
I left it unset because it always seems possible, even if very unlikely, that a generic message may not be accurate. Working with only 10 made up updates may have clouded my vision. I suppose if someone hits 100 entries on a real cluster they would certainly had had to do a z-stream update, well unless they had 95 partials or something crazy. I think some more generic message would do but agree with your general point that we should always have a reason.

Reply:
yeah, my generic message was just getting at "if you see this, please file a bug, because we always want something more specific"